ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -769,80 +771,91 @@ static Set<RangeAction<?>> getRangeActionsExcludedFromSecondPreventive(Crac crac
         Map<RangeAction<?>, Set<RangeAction<?>>> correspondanceMap = new HashMap<>();
         crac.getRangeActions().stream().filter(ra -> isRangeActionPreventive(ra, crac) && !isRangeActionAutoOrCurative(ra, crac)).forEach(pra -> {
             Set<NetworkElement> praNetworkElements = pra.getNetworkElements();
             rangeActionsNotPreventive.forEach(cra -> {
                 if (cra.getNetworkElements().equals(praNetworkElements)) {
-                    correspondanceMap.putIfAbsent(pra, new HashSet<>(Set.of(cra)));
+                    correspondanceMap.putIfAbsent(pra, new HashSet<>());
                     correspondanceMap.get(pra).add(cra);
                 }
             });
         });
-        Map<RangeAction<?>, Map<State, Set<Double>>> setPointResults = new HashMap<>();
+        Map<RangeAction<?>, Map<State, Double>> setPointResults = new HashMap<>();
         correspondanceMap.forEach((pra, associatedCras) -> {
-            setPointResults.put(pra, new HashMap<>(Map.of(preventiveState, Set.of(firstPreventiveResult.getOptimizedSetpoint(pra, preventiveState)))));
+            setPointResults.put(pra, new HashMap<>(Map.of(preventiveState, firstPreventiveResult.getOptimizedSetpoint(pra, preventiveState))));
             associatedCras.forEach(cra -> contingencyResults.forEach((state, result) -> {
                 if (isRangeActionAvailableInState(cra, state, crac)) {
-                    Map<State, Set<Double>> praResults = setPointResults.get(pra);
-                    double craSetPoint = result.getOptimizedSetpoint(cra, state);
-                    praResults.putIfAbsent(state, new HashSet<>(Set.of(craSetPoint)));
-                    praResults.get(state).add(craSetPoint);
+                    setPointResults.get(pra).putIfAbsent(state, result.getOptimizedSetpoint(cra, state));
                 }
             }));
         });
         multipleInstantRangeActions.forEach(ra -> {
-            setPointResults.put(ra, new HashMap<>(Map.of(preventiveState, Set.of(firstPreventiveResult.getOptimizedSetpoint(ra, preventiveState)))));
+            setPointResults.put(ra, new HashMap<>(Map.of(preventiveState, firstPreventiveResult.getOptimizedSetpoint(ra, preventiveState))));
             contingencyResults.forEach((state, result) -> {
                 if (isRangeActionAvailableInState(ra, state, crac)) {
-                    setPointResults.get(ra).put(state, Set.of(result.getOptimizedSetpoint(ra, state)));
+                    setPointResults.get(ra).put(state, result.getOptimizedSetpoint(ra, state));
                 }
             });
         });
         return setPointResults;
     }
 
+    /**
+     * Checks if raUsageLimits are at risk if we choose to re-optimize a range action.
+     * Returns True if it's at risk, False otherwise.
+     */
+    private static boolean shouldRemoveRaDueToUsageLimits(Crac crac, State state, RangeAction<?> rangeAction, Set<NetworkAction> activatedNetworkActions) {
```

Comment:
It seems inefficient to check for each ra, for each state if a ra should be removed, because you have to recompute the ra limits of the given state for each ra.
If you looped through the states and checked if for each one the ra should be removed you might be able to mutualize some computations.

Reply:
Indeed.
I changed the structure of the setPointResults Map to loop through states.